Rumer report

lomamech opened the pull request:
- Implement the rumer report
Note: This PR requires the execution of updates in the migrate file

closes #5395

lomamech committed:
feature(Implement rumer report)
- Implement the rumer report
Note: This PR requires the execution of updates in the migrate file

closes #5395

lomamech committed:
Clean and sanitaze code

lomamech committed:
improvement(Rumer Report)
- Use table stock_movement_status instead stock_movement

lomamech committed:
Using the name of Month and Year instead date from and date end on Header of Rumer Report

jniles commented on server/controllers/stock/reports/rumer.report.handlebars:
```diff
@@ -0,0 +1,54 @@
+            {{#each configurationData}}
+            <tr>
+              <td> {{ inventoryText }} </td>
+              <td><strong> {{ quantityOpening }} </strong></td>
```
These should be `class="text-right"` as well.

jniles commented on server/controllers/stock/reports/rumer.report.handlebars:
```diff
@@ -0,0 +1,54 @@
+                </td>
+              {{/each}}
+              <td><strong> {{ quantityTotalExit }} </strong></td>
+              <td><strong> {{ quantityEnding }} </strong></td>  
```
Both `quantityTotalExit` and `quantityEnding` should be `text-right` as well.

jniles commented on server/controllers/stock/reports/stock/rumer.js:
```diff
@@ -0,0 +1,172 @@
+
+    const sqlDailyConsumption = `
+      SELECT BUID(sms.inventory_uuid) AS uuid, inv.code, inv.text AS inventory_text,
+      sms.out_quantity_consumption AS quantity, sms.date
```
I talked to AG Neron, and apparently this needs to have both the consumption and the exits due to loss.  So, this should be changed to
```sql
SELECT (sms.out_quantity_consumption + sms.out_quantity_exit) AS quantity
```

jniles commented on server/controllers/stock/reports/stock/rumer.js:
```diff
@@ -0,0 +1,172 @@
+      sms.out_quantity_consumption AS quantity, sms.date
+        FROM stock_movement_status AS sms
+        JOIN inventory AS inv ON inv.uuid = sms.inventory_uuid
+      WHERE sms.depot_uuid = ? AND DATE(sms.date) >= DATE(?) AND DATE(sms.date) <= DATE(?)
```
Because [the `sms.date` column is a `DATE`](https://github.com/IMA-WorldHealth/bhima/blob/97eabad91d26934b575744c0bc27f3067b9e376b/server/models/schema.sql#L2027) and not a `DATETIME` or` TIMESTAMP`, you don't need to wrap it in `DATE()`.   By wrapping it, you lose a performance gain from the INDEX.

You can change this code to:
```sql
WHERE sms.depot_uuid = ? AND sms.date >= DATE(?) AND sms.date <= DATE(?)
```

jniles commented on server/controllers/stock/reports/stock/rumer.js:
```diff
@@ -0,0 +1,172 @@
+      SUM(sms.out_quantity_consumption) AS quantity, sms.date
+        FROM stock_movement_status AS sms
+        JOIN inventory AS inv ON inv.uuid = sms.inventory_uuid
+      WHERE sms.depot_uuid = ? AND DATE(sms.date) >= DATE(?) AND DATE(sms.date) <= DATE(?)
```
Same comment about `DATE` here and the `exit` + `consumption` calculation.

jniles commented on server/controllers/stock/reports/stock/rumer.js:
```diff
@@ -0,0 +1,172 @@
+    const sqlStockEntryMonth = `
+      SELECT BUID(inv.uuid) AS uuid, inv.code, inv.text AS inventory_text, SUM(sm.quantity) AS quantity,
+      sm.date, sm.is_exit
+        FROM stock_movement AS sm
```
You should be able to get this from `stock_movement_status` from the column [`in_quantity`](https://github.com/IMA-WorldHealth/bhima/blob/97eabad91d26934b575744c0bc27f3067b9e376b/server/models/schema.sql#L2029).

jniles commented on server/controllers/stock/reports/stock/rumer.js:
```diff
@@ -0,0 +1,172 @@
+      core.getInventoryQuantityAndConsumption(parameterEndingStock),
+    ]);
+
+    inventoriesEnding.forEach(inventory => {
```
It might be helpful to add an option to remove the inventories with `0` movements.  Could you add that?

lomamech replied:
A small precision, do you allude to the products whose quantity at the beginning of the month is null and have had no movement during the month

jniles replied:
Yes, that is correct.

jniles replied:
Basically allow the user to remove products where the beginning balance is 0, the movements are 0, and the ending balance are 0.

jniles commented on server/controllers/stock/reports/rumer.report.handlebars:
```diff
@@ -0,0 +1,54 @@
+            {{#each header as | key |}}
+              <th style="min-width:20px;" class="text-center">{{key}}</th>
+            {{/each}}
+            <th style="width: 7.5%;" class="text-center">{{translate 'FORM.LABELS.TOTAL'}} </th>
```
Could you rename this column to "Total Exits"?  It would match the "Total Entries" at the beginning a lot better.


![image](https://user-images.githubusercontent.com/896472/117121341-5c453c00-ad8c-11eb-83b3-a16c1c6940e8.png)

jniles commented on server/controllers/stock/reports/stock/rumer.js:
```diff
@@ -0,0 +1,172 @@
+      inventory.dailyConsumption = dailyConsumption;
+    });
+
+    if (inventoriesOpening.length) {
```
You are doing lots of these loops that look very similar.

Can you combine them into a single loop?  Instead of checking if `inventoriesOpening.length`, then doing a `configurationData.forEach()`, just add the `inventoriesOpening.forEach()` into the `configurationData.forEach()` loop.  Like this:

```js
configurationData.forEach(row => {

  // first, add the opening balances:
  inventoriesOpening.forEach(inv => {
    if (row.inventoryUuid === inv.inventory_uuid) { row.quantityOpening = inv.quantity; }
  });

  // next process the entries
  inventoriesEntry.forEach(inv => {
    if (row.inventoryUuid === inv.uuid) { row.quantityTotalEntry = inv.quantity; }
  });

  // finally process the exit quantities
  monthlyConsumption.forEach(inv => {
    if (row.inventoryUuid === inv.uuid) { row.quantityTotalExit = inv.quantity; }
  });
});
```

jniles reviewed (commented):
@lomamech just a few comments here.  Overall looks really good!

I think my inline comments should be clear enough.  Let me know if you have any questions.

lomamech committed:
improvement(Rumer Report)
- Remove inventories out of stock and 0 movement

lomamech committed:
Add option for condensed report

jniles commented on client/src/modules/reports/generate/rumer_report/rumer_report.config.js:
```diff
@@ -0,0 +1,86 @@
+
+  // default values
+  vm.reportDetails = {
+    includePurchaseEntry : 1,
```
I think we can probably remove this. I don't think it is ever used.

jniles commented on server/controllers/stock/reports/stock/rumer.js:
```diff
@@ -0,0 +1,212 @@
+      GROUP BY inv.uuid;
+    `;
+
+    const sqlStockEntryMonth = `
```
It looks like all these use the same query.  I think we can probably gain a significant performance boost by combining them.  I'll go ahead and make that change in a commit.

jniles committed:
chore(report): condense RUMER queries

Avoids repeating too many SQL queries and just uses the same one for all
the RUMER fields.  Also adds red text when the ending stock is negative.

jniles reviewed (commented):
Everything here looks good to me.  I think I see an easy performance improvement, so I'll make that change in a commit myself.  After that, this is ready to roll.  Great job @lomamech!
